packages/core/src/events/input/load/vectorstore/base.ts

```diff
@@ -0,0 +1,233 @@
+import type { Embeddings } from "../../../embeddings/base.js";
+import type { Context } from "../docs/context.js";
+import { BaseRetriever, BaseRetrieverInput } from "./baseretriever.js";
+import { Callable, CallableConfig} from "../../../../record/callable.js";
+
+type AddContextOptions = Record<string, any>;
+
+export type MaxMarginalRelevanceSearchOptions<FilterType> = {
+  k: number;
+  fetchK?: number;
+  lambda?: number;
+  filter?: FilterType;
+};
+
+
+export type VectorStoreRetrieverMMRSearchKwargs = {
+  fetchK?: number;
+  lambda?: number;
+};
+
+
+export type VectorStoreRetrieverInput<V extends VectorStore> =
+  BaseRetrieverInput &
+    (
+      | {
+          vectorStore: V;
+          k?: number;
+          filter?: V["FilterType"];
+          searchType?: "similarity";
+        }
+      | {
+          vectorStore: V;
+          k?: number;
+          filter?: V["FilterType"];
+          searchType: "mmr";
+          searchKwargs?: VectorStoreRetrieverMMRSearchKwargs;
+        }
+    );
+
+
+export class VectorStoreRetriever<V extends VectorStore = VectorStore> extends BaseRetriever {
+    _namespace: string[];
+  
+    vectorStore: V;
+  
+    k = 4;
+  
+    searchType = "similarity";
+  
+    searchKwargs?: VectorStoreRetrieverMMRSearchKwargs;
+  
+    filter?: V["FilterType"];
+  
+    _vectorstoreType(): string {
+      return this.vectorStore._vectorstoreType();
+    }
+  
+    constructor(fields: VectorStoreRetrieverInput<V>) {
+      super(fields);
+      this.vectorStore = fields.vectorStore;
+      this.k = fields.k ?? this.k;
+      this.searchType = fields.searchType ?? this.searchType;
+      this.filter = fields.filter;
+      if (fields.searchType === "mmr") {
+        this.searchKwargs = fields.searchKwargs;
+      }
+    }
+  
+    async _getRelevantContexts(
+      query: string,
+    ): Promise<Context[]> {
+      if (this.searchType === "mmr") {
+        if (typeof this.vectorStore.maxMarginalRelevanceSearch !== "function") {
+          throw new Error(
+            `The vector store backing this retriever, ${this._vectorstoreType()} does not support max marginal relevance search.`
+          );
+        }
+        return this.vectorStore.maxMarginalRelevanceSearch(query,{k: this.k,filter: this.filter,...this.searchKwargs}, undefined);
+      }
+      return this.vectorStore.similaritySearch(
+        query,
+        this.k,
+        this.filter,
+      );
+    }
+  
+    async addContexts(
+      documents: Context[],
+      options?: AddContextOptions
+    ): Promise<string[] | void> {
+      return this.vectorStore.addContexts(documents, options);
+    }
+}
+
+export abstract class VectorStore extends Callable {
+  declare FilterType: object | string;
+
+  lc_namespace = ["Encre", "vectorstores", this._vectorstoreType()];
+
+  embeddings: Embeddings;
+
+  constructor(embeddings: Embeddings, dbConfig: Record<string, any>) {
```

Comment:
The input of your Callable constructor should always be the following formating: `constructor(fields?: Partial<XXXParams>)`, where `XXXParams` can be a type or interface of your inputs. This will guarantee the input format to be JSON-like.

Reply:
would XXXParams require specific formats?

Reply:
The Params is a JSON object. I think in your case the Params should wrap the `embeddings` and `dbConfig` but I'm not quite sure if the `dbConfig` is something user should modify whenever they invoke the event. If it is something that require updates for every `invoke()`, then this config should be included in the `CallOptions`

Reply:
Also check [constructor in Serializable](https://github.com/VictorS67/encre/blob/416f2af6395907a975fcb1f0b0ae011d0229634e/packages/core/src/load/serializable.ts#L133) for the argument passed. You can see the argument should be wrapped up in `object` type, which is a JSON object